source/extensions/transport_sockets/tls/ocsp/ocsp.cc

```diff
@@ -0,0 +1,307 @@
+#include "extensions/transport_sockets/tls/ocsp/ocsp.h"
+
+#include "common/common/utility.h"
+
+#include "extensions/transport_sockets/tls/ocsp/asn1_utility.h"
+#include "extensions/transport_sockets/tls/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Tls {
+namespace Ocsp {
+
+namespace CertUtility = Envoy::Extensions::TransportSockets::Tls::Utility;
+
+static const unsigned CONS = CBS_ASN1_CONSTRUCTED;
+static const unsigned CONT = CBS_ASN1_CONTEXT_SPECIFIC;
+
+namespace {
+
+unsigned parseTag(CBS& cbs) {
+  unsigned tag;
+  if (!CBS_get_any_asn1_element(&cbs, nullptr, &tag, nullptr)) {
+    throw EnvoyException("Failed to parse ASN.1 element tag");
+  }
+  return tag;
+}
+
+std::unique_ptr<OcspResponse> readDerEncodedOcspResponse(const std::string& der) {
+  CBS cbs;
+  CBS_init(&cbs, reinterpret_cast<const uint8_t*>(der.c_str()), der.size());
+
+  auto resp = Asn1OcspUtility::parseOcspResponse(cbs);
+  if (CBS_len(&cbs) != 0) {
+    throw EnvoyException("Data contained more than a single OCSP response");
+  }
+
+  return resp;
+}
+
+void skipResponderId(CBS& cbs) {
+  // ResponderID ::= CHOICE {
+  //    byName               [1] Name,
+  //    byKey                [2] KeyHash
+  // }
+  //
+  // KeyHash ::= OCTET STRING -- SHA-1 hash of responder's public key
+  //    (excluding the tag and length fields)
+
+  if (Asn1Utility::isOptionalPresent(cbs, nullptr, CONS | CONT | 1) ||
+      Asn1Utility::isOptionalPresent(cbs, nullptr, CONS | CONT | 2)) {
+    return;
+  }
+
+  throw EnvoyException(absl::StrCat("Unknown choice for Responder ID: ", parseTag(cbs)));
+}
+
+} // namespace
+
+OcspResponse::OcspResponse(OcspResponseStatus status, ResponsePtr&& response)
```

Comment:
Possibly take `ResponsePtr` instead of `ResponsePtr&&`, since you unconditionally take ownership. (I can't remember all the nuances of `&&`, so maybe that's not possible).